syntax = "proto3";

package DistributionSystem;

// Сервис выполнения задач
service TaskExecutionService {
  rpc DistributeDetectionTask(TaskInfoRequest) returns (PathResponse);
  rpc ExecuteDetectionTask(stream ImageRequest) returns (stream ImageResponse);
}

message TaskInfoRequest {
  string task_data = 1;
}

message PathResponse {
  string path = 1;
}

message ImageRequest {
  string client_name = 1;
  string filename = 2;
  string path_to_destination_node = 3;
  bool is_last_chunk_of_photo = 4;
  bool is_last_photo = 5;
  bytes data = 6;
}

message ImageResponse {
  string filename = 1;
  bool is_last_chunk_of_photo = 2;
  bool is_last_photo = 3;
  bytes data = 4;
}


// Сервис для распределения тасок по узлам системы
service DistributionTasksService {
  rpc CollectData(CollectDataRequest) returns (CollectedData);
}

message CollectDataRequest {
  string to_server_address = 1;
}

message CollectedData{
  string server_address = 1;
  string collected_info = 2;
}


// Сервис для достижения отказоустойчивости в системе
service FaultToleranceService {
  rpc Ping(PingRequest) returns (PingResponse);
}

message PingRequest {
  string to_server_address = 1;
}

message PingResponse {
  string server_address = 1;
  string is_alive = 2;
}


// Сервис для масштабируемости системы:
service ScalabilityService {
  rpc GetTopology(TopologyRequest) returns (TopologyResponse);
}

message TopologyRequest {
  string to_server_address = 1;
}

message TopologyResponse {
  string parent_address = 1;
  string children_addresses = 2;
}
